File: Orb_Assistant/orbskin-scaffold-v0.1/orbskin/loaders/web/skin_loader.py
"""
ORB SKIN LOADER — WEB ORB (FastAPI)
Handles .orbskin packages for the website ORB / Orb Weaver backend.

Endpoints:
  POST   /skin/load          Upload and activate a .orbskin package
  POST   /skin/rollback      Rollback to previous skin
  GET    /skin/active        Get current active skin info
  DELETE /skin/clear         Clear active skin
  GET    /skin/asset/{path}  Serve an extracted skin asset

Wiring: Include this router in your main FastAPI app:
  from skin_loader import router as skin_router
  app.include_router(skin_router, prefix="/api")
"""

import hashlib
import io
import json
import os
import shutil
import zipfile
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, HTTPException, UploadFile, File
from fastapi.responses import FileResponse, JSONResponse
from pydantic import BaseModel

# ─────────────────────────────────────────────
# CONFIG
# ─────────────────────────────────────────────

SKIN_STORE_DIR = Path(os.getenv("ORB_SKIN_STORE", "./data/skins"))
RUNTIME_VERSION = "1.0.0"
ORB_TARGET = "website"

router = APIRouter(tags=["OrbSkin"])

# ─────────────────────────────────────────────
# IN-MEMORY STATE  (swap for DB/Redis in production)
# ─────────────────────────────────────────────

_state: dict = {
    "active": None,       # SkinBundle dict
    "rollback": None,     # SkinBundle dict
    "active_path": None,
    "rollback_path": None,
}

# ─────────────────────────────────────────────
# MODELS
# ─────────────────────────────────────────────

class ValidationIssue(BaseModel):
    code: str
    field: Optional[str] = None
    message: str

class ValidationResult(BaseModel):
    valid: bool
    skin_id: str
    errors: list[ValidationIssue]
    warnings: list[ValidationIssue]
    validated_at: str

class SkinInfo(BaseModel):
    skin_id: str
    name: str
    loaded_at: str
    rollback_available: bool

# ─────────────────────────────────────────────
# VALIDATION  (mirrors TS validator logic)
# ─────────────────────────────────────────────

def validate_manifest(manifest: dict, for_target: str, package_hash: str) -> ValidationResult:
    errors: list[ValidationIssue] = []
    warnings: list[ValidationIssue] = []
    now = datetime.now(timezone.utc).isoformat()

    # Schema version
    if manifest.get("schema_version") != "1.0":
        warnings.append(ValidationIssue(
            code="SCHEMA_VERSION_MISMATCH",
            field="schema_version",
            message=f"Expected 1.0, got {manifest.get('schema_version')} — may still load"
        ))

    # Required fields
    for field in ["skin_id", "name", "version"]:
        if not manifest.get(field):
            errors.append(ValidationIssue(code="MISSING_FIELD", field=field, message=f'"{field}" required'))

    # Creator
    creator = manifest.get("creator", {})
    if not creator.get("creator_id") or not creator.get("display_name"):
        errors.append(ValidationIssue(code="MISSING_FIELD", field="creator", message="creator.creator_id and creator.display_name required"))

    # Target compatibility
    cls = manifest.get("classification", {})
    supported = cls.get("supported_orbs", [])
    if supported and for_target not in supported and "all" not in supported:
        errors.append(ValidationIssue(
            code="UNSUPPORTED_TARGET",
            field="classification.supported_orbs",
            message=f'Skin does not support target "{for_target}". Supported: {", ".join(supported)}'
        ))

    # Visuals
    vis = manifest.get("visuals", {})
    if not vis:
        errors.append(ValidationIssue(code="MISSING_FIELD", field="visuals", message='"visuals" block required'))
    else:
        for f in ["preview", "body_asset", "docked_icon"]:
            if not vis.get(f):
                errors.append(ValidationIssue(code="MISSING_ASSET", field=f"visuals.{f}", message=f'"visuals.{f}" required'))

    # Behavior hard walls
    bl = manifest.get("behavior_limits", {})
    if not bl:
        errors.append(ValidationIssue(code="MISSING_FIELD", field="behavior_limits", message='"behavior_limits" block required'))
    else:
        if bl.get("changes_visuals_only") is not True:
            errors.append(ValidationIssue(
                code="BEHAVIOR_VIOLATION",
                field="behavior_limits.changes_visuals_only",
                message="changes_visuals_only must be true"
            ))
        for hard_false in ["may_add_permissions", "may_add_tools", "may_add_network_access", "may_add_llm_access"]:
            if bl.get(hard_false) is not False:
                errors.append(ValidationIssue(
                    code="BEHAVIOR_VIOLATION",
                    field=f"behavior_limits.{hard_false}",
                    message=f"{hard_false} must be false"
                ))

    # Rights / license expiry
    rights = manifest.get("rights", {})
    if rights.get("expiry_date"):
        try:
            expiry = datetime.fromisoformat(rights["expiry_date"])
            if expiry < datetime.now(timezone.utc):
                errors.append(ValidationIssue(
                    code="LICENSE_EXPIRED",
                    field="rights.expiry_date",
                    message=f'License expired: {rights["expiry_date"]}'
                ))
        except ValueError:
            warnings.append(ValidationIssue(code="INVALID_DATE", field="rights.expiry_date", message="Could not parse expiry_date"))

    # Hash check
    integ = manifest.get("integrity", {})
    if integ:
        stored_hash = integ.get("package_hash", "").lstrip("sha256:")
        actual_hash = package_hash.lstrip("sha256:")
        if stored_hash and actual_hash and stored_hash != actual_hash:
            errors.append(ValidationIssue(
                code="HASH_MISMATCH",
                field="integrity.package_hash",
                message=f"Hash mismatch. Stored: {stored_hash[:12]}... Actual: {actual_hash[:12]}..."
            ))

    return ValidationResult(
        valid=len(errors) == 0,
        skin_id=manifest.get("skin_id", "unknown"),
        errors=errors,
        warnings=warnings,
        validated_at=now
    )

# ─────────────────────────────────────────────
# PACKAGE PROCESSING
# ─────────────────────────────────────────────

def compute_sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()

def extract_skin_package(raw_bytes: bytes, skin_id: str) -> Path:
    """Extract .orbskin zip to SKIN_STORE_DIR/<skin_id>/"""
    dest_dir = SKIN_STORE_DIR / skin_id
    if dest_dir.exists():
        shutil.rmtree(dest_dir)
    dest_dir.mkdir(parents=True, exist_ok=True)

    with zipfile.ZipFile(io.BytesIO(raw_bytes)) as zf:
        for name in zf.namelist():
            zf.extract(name, dest_dir)

    return dest_dir

def build_asset_url(skin_id: str, filename: str) -> str:
    """Build the URL the React renderer will use to fetch this asset."""
    # Served by the GET /skin/asset/{path} endpoint below
    return f"/api/skin/asset/{skin_id}/{filename}"

def build_bundle(manifest: dict, skin_id: str) -> dict:
    vis = manifest.get("visuals", {})

    animations = {
        anim: build_asset_url(skin_id, f"animations/{anim}")
        for anim in vis.get("animations", [])
    }
    sounds = {
        snd: build_asset_url(skin_id, f"sounds/{snd}")
        for snd in vis.get("sounds", [])
    }

    return {
        "skin_id": skin_id,
        "name": manifest.get("name", ""),
        "manifest": manifest,
        "urls": {
            "preview": build_asset_url(skin_id, vis.get("preview", "")),
            "body_asset": build_asset_url(skin_id, vis.get("body_asset", "")),
            "docked_icon": build_asset_url(skin_id, vis.get("docked_icon", "")),
            "animations": animations,
            "particle_profile": build_asset_url(skin_id, vis["particle_profile"]) if vis.get("particle_profile") else None,
            "sounds": sounds,
        },
        "theme_tokens": vis.get("theme_tokens", {}),
        "loaded_at": datetime.now(timezone.utc).isoformat(),
    }

# ─────────────────────────────────────────────
# ENDPOINTS
# ─────────────────────────────────────────────

@router.post("/skin/load")
async def load_skin(file: UploadFile = File(...)):
    """Upload and activate a .orbskin package."""
    if not file.filename or not file.filename.endswith(".orbskin"):
        raise HTTPException(status_code=400, detail="File must have .orbskin extension")

    raw_bytes = await file.read()
    package_hash = compute_sha256(raw_bytes)

    # Read and parse manifest
    try:
        with zipfile.ZipFile(io.BytesIO(raw_bytes)) as zf:
            if "manifest.json" not in zf.namelist():
                raise HTTPException(status_code=400, detail="manifest.json not found in package")
            manifest_text = zf.read("manifest.json").decode("utf-8")
            manifest = json.loads(manifest_text)
    except zipfile.BadZipFile:
        raise HTTPException(status_code=400, detail="Invalid zip/orbskin package")
    except json.JSONDecodeError as e:
        raise HTTPException(status_code=400, detail=f"Invalid manifest JSON: {e}")

    # Validate
    validation = validate_manifest(manifest, ORB_TARGET, package_hash)
    if not validation.valid:
        return JSONResponse(
            status_code=422,
            content={
                "ok": False,
                "error": "Validation failed",
                "validation": validation.model_dump()
            }
        )

    skin_id = manifest["skin_id"]

    # Extract
    try:
        extract_skin_package(raw_bytes, skin_id)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Extraction failed: {e}")

    # Build bundle
    bundle = build_bundle(manifest, skin_id)

    # Store rollback
    _state["rollback"] = _state["active"]
    _state["active"] = bundle

    return {"ok": True, "bundle": bundle, "validation": validation.model_dump()}


@router.post("/skin/rollback")
async def rollback_skin():
    """Rollback to the previous skin."""
    if not _state["rollback"]:
        raise HTTPException(status_code=404, detail="No rollback skin available")

    current = _state["active"]
    _state["active"] = _state["rollback"]
    _state["rollback"] = current

    return {"ok": True, "bundle": _state["active"]}


@router.get("/skin/active")
async def get_active_skin():
    """Get current active skin info."""
    if not _state["active"]:
        return {"active": None, "rollback_available": False}

    b = _state["active"]
    return {
        "active": {
            "skin_id": b["skin_id"],
            "name": b["name"],
            "loaded_at": b["loaded_at"],
        },
        "rollback_available": _state["rollback"] is not None
    }


@router.get("/skin/bundle")
async def get_active_bundle():
    """Get the full active skin bundle (for React renderer to initialize from)."""
    if not _state["active"]:
        return {"bundle": None}
    return {"bundle": _state["active"]}


@router.delete("/skin/clear")
async def clear_skin():
    """Remove the active skin. ORB returns to default appearance."""
    _state["rollback"] = _state["active"]
    _state["active"] = None
    return {"ok": True}


@router.get("/skin/asset/{skin_id}/{file_path:path}")
async def serve_skin_asset(skin_id: str, file_path: str):
    """Serve an extracted skin asset file."""
    # Security: sanitize paths to prevent traversal
    safe_id = "".join(c for c in skin_id if c.isalnum() or c in "-_")
    asset_path = SKIN_STORE_DIR / safe_id / file_path

    # Resolve and verify it stays inside the skin dir
    try:
        resolved = asset_path.resolve()
        store_resolved = SKIN_STORE_DIR.resolve()
        if not str(resolved).startswith(str(store_resolved)):
            raise HTTPException(status_code=403, detail="Access denied")
    except Exception:
        raise HTTPException(status_code=403, detail="Invalid path")

    if not resolved.exists():
        raise HTTPException(status_code=404, detail=f"Asset not found: {file_path}")

    return FileResponse(resolved)

File: Orb_Assistant/orbskin-scaffold-v0.1/orbskin/loaders/tauri/skin_loader.rs
// ORB SKIN LOADER — TAURI (Rust core)
// File: src-tauri/src/skin_loader.rs
//
// Cargo.toml additions needed:
//   [dependencies]
//   zip = "0.6"
//   sha2 = "0.10"
//   serde = { version = "1", features = ["derive"] }
//   serde_json = "1"
//   tauri = { version = "2", features = ["protocol-asset"] }
//   hex = "0.4"
//
// Registers a Tauri command: load_orb_skin(path: String) -> Result<SkinBundle, String>
// Assets served via asset:// custom protocol or local file path

use std::collections::HashMap;
use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use tauri::{AppHandle, Emitter, Manager, State};
use zip::ZipArchive;

// ─────────────────────────────────────────────
// STATE  (stored in Tauri managed state)
// ─────────────────────────────────────────────

pub struct SkinState {
    pub active: Option<SkinBundle>,
    pub rollback: Option<SkinBundle>,
    pub active_path: Option<PathBuf>,
    pub rollback_path: Option<PathBuf>,
}

impl Default for SkinState {
    fn default() -> Self {
        Self {
            active: None,
            rollback: None,
            active_path: None,
            rollback_path: None,
        }
    }
}

pub type ManagedSkinState = Mutex<SkinState>;

// ─────────────────────────────────────────────
// MANIFEST TYPES  (mirrors TS types)
// ─────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OrbSkinManifest {
    pub schema_version: String,
    pub skin_id: String,
    pub name: String,
    pub version: String,
    pub description: Option<String>,
    pub creator: ManifestCreator,
    pub classification: ManifestClassification,
    pub visuals: ManifestVisuals,
    pub behavior_limits: ManifestBehaviorLimits,
    pub rights: ManifestRights,
    pub integrity: ManifestIntegrity,
    // Optional sections
    pub marketplace: Option<serde_json::Value>,
    pub collectible: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestCreator {
    pub creator_id: String,
    pub display_name: String,
    pub verified: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestClassification {
    pub tier: String,
    pub edition_type: String,
    pub supported_orbs: Vec<String>,
    pub commercial_use: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestVisuals {
    pub preview: String,
    pub body_asset: String,
    pub docked_icon: String,
    pub animations: Vec<String>,
    pub particle_profile: Option<String>,
    pub sounds: Option<Vec<String>>,
    pub theme_tokens: Option<HashMap<String, String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestBehaviorLimits {
    pub changes_visuals_only: bool,
    pub may_change_voice_style: bool,
    pub may_change_personality_language: bool,
    pub may_add_permissions: bool,
    pub may_add_tools: bool,
    pub may_add_network_access: bool,
    pub may_add_llm_access: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestRights {
    pub license_type: String,
    pub transferable: bool,
    pub resellable: bool,
    pub max_active_orbs: u32,
    pub expiry_date: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManifestIntegrity {
    pub package_hash: String,
    pub manifest_hash: String,
    pub publisher_signature: String,
    pub signed_at: String,
    pub runtime_min_version: String,
    pub runtime_max_version: Option<String>,
}

// ─────────────────────────────────────────────
// BUNDLE  (what we emit to the webview)
// ─────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkinBundle {
    pub skin_id: String,
    pub name: String,
    pub manifest: OrbSkinManifest,
    pub urls: SkinUrls,
    pub theme_tokens: HashMap<String, String>,
    pub loaded_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkinUrls {
    pub preview: String,
    pub body_asset: String,
    pub docked_icon: String,
    pub animations: HashMap<String, String>,
    pub particle_profile: Option<String>,
    pub sounds: HashMap<String, String>,
}

// ─────────────────────────────────────────────
// VALIDATION
// ─────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize)]
pub struct ValidationResult {
    pub valid: bool,
    pub skin_id: String,
    pub errors: Vec<ValidationIssue>,
    pub warnings: Vec<ValidationIssue>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ValidationIssue {
    pub code: String,
    pub field: Option<String>,
    pub message: String,
}

fn validate_manifest(
    manifest: &OrbSkinManifest,
    for_target: &str,
    package_hash: &str,
) -> ValidationResult {
    let mut errors: Vec<ValidationIssue> = Vec::new();
    let warnings: Vec<ValidationIssue> = Vec::new();

    // Schema version
    if manifest.schema_version != "1.0" {
        errors.push(ValidationIssue {
            code: "SCHEMA_VERSION_MISMATCH".into(),
            field: Some("schema_version".into()),
            message: format!("Expected 1.0, got {}", manifest.schema_version),
        });
    }

    // Target compatibility
    let supported = &manifest.classification.supported_orbs;
    if !supported.contains(&for_target.to_string()) && !supported.contains(&"all".to_string()) {
        errors.push(ValidationIssue {
            code: "UNSUPPORTED_TARGET".into(),
            field: Some("classification.supported_orbs".into()),
            message: format!(
                "Skin does not support target \"{}\". Supported: {}",
                for_target,
                supported.join(", ")
            ),
        });
    }

    // Behavior hard walls — these are non-negotiable
    let bl = &manifest.behavior_limits;
    if !bl.changes_visuals_only {
        errors.push(ValidationIssue {
            code: "BEHAVIOR_VIOLATION".into(),
            field: Some("behavior_limits.changes_visuals_only".into()),
            message: "changes_visuals_only must be true".into(),
        });
    }
    for (field, value) in [
        ("may_add_permissions", bl.may_add_permissions),
        ("may_add_tools", bl.may_add_tools),
        ("may_add_network_access", bl.may_add_network_access),
        ("may_add_llm_access", bl.may_add_llm_access),
    ] {
        if value {
            errors.push(ValidationIssue {
                code: "BEHAVIOR_VIOLATION".into(),
                field: Some(format!("behavior_limits.{}", field)),
                message: format!("{} must be false", field),
            });
        }
    }

    // Hash check
    let stored_hash = manifest.integrity.package_hash
        .trim_start_matches("sha256:")
        .to_lowercase();
    let actual_hash = package_hash
        .trim_start_matches("sha256:")
        .to_lowercase();
    if stored_hash != actual_hash {
        errors.push(ValidationIssue {
            code: "HASH_MISMATCH".into(),
            field: Some("integrity.package_hash".into()),
            message: format!(
                "Hash mismatch. Stored: {}... Actual: {}...",
                &stored_hash[..12.min(stored_hash.len())],
                &actual_hash[..12.min(actual_hash.len())]
            ),
        });
    }

    ValidationResult {
        valid: errors.is_empty(),
        skin_id: manifest.skin_id.clone(),
        errors,
        warnings,
    }
}

// ─────────────────────────────────────────────
// TAURI COMMANDS
// ─────────────────────────────────────────────

/// Load a .orbskin file. Call from JS: invoke("load_orb_skin", { path: "..." })
#[tauri::command]
pub async fn load_orb_skin(
    app: AppHandle,
    state: State<'_, ManagedSkinState>,
    path: String,
) -> Result<SkinBundle, String> {
    let skin_path = PathBuf::from(&path);

    // Read bytes
    let raw = fs::read(&skin_path).map_err(|e| format!("Cannot read file: {e}"))?;

    // Compute hash
    let mut hasher = Sha256::new();
    hasher.update(&raw);
    let hash_hex = hex::encode(hasher.finalize());
    let package_hash = format!("sha256:{hash_hex}");

    // Unzip
    let cursor = std::io::Cursor::new(raw);
    let mut archive = ZipArchive::new(cursor).map_err(|e| format!("Invalid zip: {e}"))?;

    // Read manifest
    let manifest_text = {
        let mut mf = archive
            .by_name("manifest.json")
            .map_err(|_| "manifest.json not found in package")?;
        let mut buf = String::new();
        mf.read_to_string(&mut buf).map_err(|e| format!("Cannot read manifest: {e}"))?;
        buf
    };
    let manifest: OrbSkinManifest =
        serde_json::from_str(&manifest_text).map_err(|e| format!("Invalid manifest JSON: {e}"))?;

    // Validate
    let validation = validate_manifest(&manifest, "desktop", &package_hash);
    if !validation.valid {
        let error_msgs: Vec<String> = validation.errors.iter().map(|e| e.message.clone()).collect();
        return Err(format!("Skin validation failed: {}", error_msgs.join("; ")));
    }

    // Extract to app data dir
    let app_data = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("Cannot get app data dir: {e}"))?;
    let skin_dir = app_data.join("skins").join(&manifest.skin_id);
    fs::create_dir_all(&skin_dir).map_err(|e| format!("Cannot create skin dir: {e}"))?;

    // Write all files except manifest
    for i in 0..archive.len() {
        let mut file = archive.by_index(i).map_err(|e| format!("Zip error: {e}"))?;
        let name = file.name().to_string();
        if name == "manifest.json" || file.is_dir() {
            continue;
        }
        let dest = skin_dir.join(&name);
        if let Some(parent) = dest.parent() {
            fs::create_dir_all(parent).ok();
        }
        let mut bytes = Vec::new();
        file.read_to_end(&mut bytes).map_err(|e| format!("Read error: {e}"))?;
        fs::write(&dest, bytes).map_err(|e| format!("Write error: {e}"))?;
    }

    // Build URL helper — use convertFileSrc on the JS side, or return absolute paths
    // Tauri v2: frontend uses convertFileSrc() to turn paths into asset:// URLs
    let to_path = |filename: &str| -> String {
        skin_dir.join(filename).to_string_lossy().to_string()
    };

    let mut animations: HashMap<String, String> = HashMap::new();
    for anim in &manifest.visuals.animations {
        animations.insert(anim.clone(), to_path(&format!("animations/{anim}")));
    }

    let mut sounds: HashMap<String, String> = HashMap::new();
    for snd in manifest.visuals.sounds.iter().flatten() {
        sounds.insert(snd.clone(), to_path(&format!("sounds/{snd}")));
    }

    let bundle = SkinBundle {
        skin_id: manifest.skin_id.clone(),
        name: manifest.name.clone(),
        urls: SkinUrls {
            preview: to_path(&manifest.visuals.preview),
            body_asset: to_path(&manifest.visuals.body_asset),
            docked_icon: to_path(&manifest.visuals.docked_icon),
            animations,
            particle_profile: manifest
                .visuals
                .particle_profile
                .as_ref()
                .map(|p| to_path(p)),
            sounds,
        },
        theme_tokens: manifest.visuals.theme_tokens.clone().unwrap_or_default(),
        manifest,
        loaded_at: chrono::Utc::now().to_rfc3339(),
    };

    // Store rollback
    {
        let mut s = state.lock().unwrap();
        s.rollback = s.active.clone();
        s.rollback_path = s.active_path.clone();
        s.active = Some(bundle.clone());
        s.active_path = Some(skin_path);
    }

    // Emit to webview
    app.emit("skin:applied", &bundle).ok();

    Ok(bundle)
}

/// Rollback to the previous skin
#[tauri::command]
pub async fn rollback_orb_skin(
    app: AppHandle,
    state: State<'_, ManagedSkinState>,
) -> Result<SkinBundle, String> {
    let mut s = state.lock().unwrap();
    let prev = s.rollback.clone().ok_or("No rollback skin available")?;
    let current = s.active.clone();
    s.active = Some(prev.clone());
    s.rollback = current;
    app.emit("skin:applied", &prev).ok();
    Ok(prev)
}

/// Get active skin info
#[tauri::command]
pub fn get_active_skin(
    state: State<'_, ManagedSkinState>,
) -> Option<HashMap<String, String>> {
    let s = state.lock().unwrap();
    s.active.as_ref().map(|b| {
        let mut m = HashMap::new();
        m.insert("skin_id".into(), b.skin_id.clone());
        m.insert("name".into(), b.name.clone());
        m
    })
}

/// Clear active skin
#[tauri::command]
pub async fn clear_orb_skin(
    app: AppHandle,
    state: State<'_, ManagedSkinState>,
) -> Result<(), String> {
    let mut s = state.lock().unwrap();
    s.rollback = s.active.clone();
    s.active = None;
    app.emit("skin:cleared", ()).ok();
    Ok(())
}

// ─────────────────────────────────────────────
// REGISTER  — call in main.rs
// ─────────────────────────────────────────────

// In your src-tauri/src/main.rs or lib.rs builder:
//
// use crate::skin_loader::{
//     ManagedSkinState, SkinState,
//     load_orb_skin, rollback_orb_skin, get_active_skin, clear_orb_skin
// };
//
// tauri::Builder::default()
//     .manage(ManagedSkinState::new(SkinState::default()))
//     .invoke_handler(tauri::generate_handler![
//         load_orb_skin,
//         rollback_orb_skin,
//         get_active_skin,
//         clear_orb_skin,
//     ])
//     .run(tauri::generate_context!())
//     .expect("error running tauri app");
